xdp-bench/xdp-bench.8
.TH "xdp-bench" "8" "FEBRUARY 19, 2026" "V1.6.3" "A simple XDP benchmarking tool"
.SH "NAME"
XDP-bench \- a simple XDP benchmarking tool
.SH "SYNOPSIS"
.PP
XDP-bench is a benchmarking utility for exercising the different operation modes
of XDP. It is intended to be a simple program demonstrating the various
operating modes; these include dropping packets, hairpin forwarding (using the
\fIXDP_TX\fP return code), and redirection using the various in-kernel packet
redirection facilities.

.PP
The drop and TX modes support various options to control whether packet data is
touched (read or written) before being dropped or transmitted. The redirection
modes support using the simple ifindex-based \fIbpf_redirect\fP helper, the
\fIbpf_redirect_map\fP helper using a cpumap as its target, \fIbpf_redirect_map\fP using
a devmap as its target, and the devmap's broadcast mode which allows redirecting
to multiple devices.

.PP
There is more information on the meaning of the output in both default (terse)
and extended output mode, in the \fBOutput Format Description\fP section below.
.SS "Running xdp-bench"
.PP
The syntax for running xdp-bench is:

.RS
.nf
\fCUsage: xdp-bench COMMAND [options]

COMMAND can be one of:
       drop           - Drop all packets on an interface
       pass           - Pass all packets to the network stack
       tx             - Transmit packets back out on an interface (hairpin forwarding)
       redirect       - XDP redirect using the bpf_redirect() helper
       redirect-cpu   - XDP CPU redirect using BPF_MAP_TYPE_CPUMAP
       redirect-map   - XDP redirect using BPF_MAP_TYPE_DEVMAP
       redirect-multi - XDP multi-redirect using BPF_MAP_TYPE_DEVMAP and the BPF_F_BROADCAST flag
       xsk-drop       - AF_XDP socket-based drop
       xsk-tx         - AF_XDP socket-based hairpin forwarding
\fP
.fi
.RE

.PP
Each command, and its options are explained below. Or use \fIxdp\-bench COMMAND
\-\-help\fP to see the options for each command.
.SH "The DROP command"
.PP
In this mode, \fIxdp\-bench\fP installs an XDP program on an interface that simply
drops all packets. There are options to control what to do with the packet
before dropping it (touch the packet data or not), as well as which statistics
to gather. This is a basic benchmark for the baseline (best-case) performance of
XDP on an interface.

.PP
The syntax for the \fIdrop\fP command is:

.PP
\fIxdp\-bench drop [options] <ifname>\fP

.PP
Where \fI<ifname>\fP is the name of the interface the XDP program should be
installed on.

.PP
The supported options are:
.SS "-p, --packet-operation <ACTION>"
.PP
Specify which operation should be taken on the packet before dropping it. The
following actions are available:

.RS
.nf
\fCno-touch		- Drop the packet without touching the packet data
read-data		- Read a field in the packet header before dropping
parse-ip		- Parse the IP header field before dropping
swap-macs		- Swap the source and destination MAC addresses before dropping
\fP
.fi
.RE

.PP
Whether to touch the packet before dropping it can have a significant
performance impact as this requires bringing packet data into the CPU cache (and
flushing it back out if writing).

.PP
The default for this option is \fIno\-touch\fP.
.SS "-l, --load-mode <MODE>"
.PP
Specify which mechanism xdp-bench should use to load (and store) the packet data.
The following modes are available:

.RS
.nf
\fCdpa		- Use traditional Direct Packet Access from the XDP program
load-bytes	- Use the xdp_load_bytes() and xdp_store_bytes() helper functions
\fP
.fi
.RE

.PP
This can be used to benchmark the various packet access modes supported by the
kernel.

.PP
The default for this option is \fIdpa\fP.
.SS "-r, --rxq-stats"
.PP
If set, the XDP program will also gather statistics on which receive queue index
each packet was received on. This is displayed in the extended output mode along
with per-CPU data (which, depending on the hardware configuration may or may not
be equivalent).
.SS "-i, --interval <SECONDS>"
.PP
Set the polling interval for collecting all statistics and displaying them to
the output. The unit of interval is in seconds.
.SS "-e, --extended"
.PP
Start xdp-bench in "extended" output mode. If not set, xdp-bench will start in
"terse" mode. The output mode can be switched by hitting C-$\ while the program
is running. See also the \fBOutput Format Description\fP section below.
.SS "-m, --mode"
.PP
Selects the XDP program mode (native or skb). Note that native XDP mode is the
default, and loading the redirect program in skb manner is neither performant,
nor recommended. However, this option is useful if the interface driver lacks
native XDP support, or when simply testing the tool.
.SS "-v, --verbose"
.PP
Enable verbose logging. Supply twice to enable verbose logging from the
underlying \fIlibxdp\fP and \fIlibbpf\fP libraries.
.SS "--version"
.PP
Show the application version and exit.
.SS "-h, --help"
.PP
Display a summary of the available options
.SH "The PASS command"
.PP
In this mode, \fIxdp\-bench\fP installs an XDP program on an interface that passes
all packets to the network stack after processing them (returning \fIXDP_PASS\fP).
There are options to control what to do with the packet before passing it
(touch the packet data or not), as well as which statistics to gather. This is a
basic benchmark for the overhead of installing an XDP program on an interface
while still running the regular network stack.

.PP
The syntax for the \fIpass\fP command is:

.PP
\fIxdp\-bench pass [options] <ifname>\fP

.PP
Where \fI<ifname>\fP is the name of the interface the XDP program should be
installed on.

.PP
The supported options are:
.SS "-p, --packet-operation <ACTION>"
.PP
Specify which operation should be taken on the packet before passing it. The
following actions are available:

.RS
.nf
\fCno-touch		- Pass the packet without touching the packet data
read-data		- Read a field in the packet header before passing
parse-ip		- Parse the IP header field before passing
swap-macs		- Swap the source and destination MAC addresses before passing
\fP
.fi
.RE

.PP
The default for this option is \fIno\-touch\fP.
.SS "-l, --load-mode <MODE>"
.PP
Specify which mechanism xdp-bench should use to load (and store) the packet data.
The following modes are available:

.RS
.nf
\fCdpa		- Use traditional Direct Packet Access from the XDP program
load-bytes	- Use the xdp_load_bytes() and xdp_store_bytes() helper functions
\fP
.fi
.RE

.PP
This can be used to benchmark the various packet access modes supported by the
kernel.

.PP
The default for this option is \fIdpa\fP.
.SS "-r, --rxq-stats"
.PP
If set, the XDP program will also gather statistics on which receive queue index
each packet was received on. This is displayed in the extended output mode along
with per-CPU data (which, depending on the hardware configuration may or may not
be equivalent).
.SS "-i, --interval <SECONDS>"
.PP
Set the polling interval for collecting all statistics and displaying them to
the output. The unit of interval is in seconds.
.SS "-e, --extended"
.PP
Start xdp-bench in "extended" output mode. If not set, xdp-bench will start in
"terse" mode. The output mode can be switched by hitting C-$\ while the program
is running. See also the \fBOutput Format Description\fP section below.
.SS "-m, --mode"
.PP
Selects the XDP program mode (native or skb). Note that native XDP mode is the
default, and loading the redirect program in skb manner is neither performant,
nor recommended. However, this option is useful if the interface driver lacks
native XDP support, or when simply testing the tool.
.SS "-v, --verbose"
.PP
Enable verbose logging. Supply twice to enable verbose logging from the
underlying \fIlibxdp\fP and \fIlibbpf\fP libraries.
.SS "--version"
.PP
Show the application version and exit.
.SS "-h, --help"
.PP
Display a summary of the available options
.SH "The TX command"
.PP
In this mode, \fIxdp\-bench\fP installs an XDP program on an interface that performs
so-called "hairpin forwarding", which means each packet is transmitted back out
the same interface (using the \fIXDP_TX\fP return code).. There are options to
control what to do with the packet before transmitting it (touch the packet data
or not), as well as which statistics to gather.

.PP
The syntax for the \fItx\fP command is:

.PP
\fIxdp\-bench tx [options] <ifname>\fP

.PP
Where \fI<ifname>\fP is the name of the interface the XDP program should be
installed on.

.PP
The supported options are:
.SS "-p, --packet-operation <ACTION>"
.PP
Specify which operation should be taken on the packet before transmitting it. The
following actions are available:

.RS
.nf
\fCno-touch		- Transmit the packet without touching the packet data
read-data		- Read a field in the packet header before transmitting
parse-ip		- Parse the IP header field before transmitting
swap-macs		- Swap the source and destination MAC addresses before transmitting
\fP
.fi
.RE

.PP
To allow the packet to be successfully transmitted back to the sender, the MAC
addresses have to be swapped, so that the source MAC matches the network device.
However, there is a performance overhead in doing swapping, so this option
allows this function to be turned off.

.PP
The default for this option is \fIswap\-macs\fP.
.SS "-l, --load-mode <MODE>"
.PP
Specify which mechanism xdp-bench should use to load (and store) the packet data.
The following modes are available:

.RS
.nf
\fCdpa		- Use traditional Direct Packet Access from the XDP program
load-bytes	- Use the xdp_load_bytes() and xdp_store_bytes() helper functions
\fP
.fi
.RE

.PP
This can be used to benchmark the various packet access modes supported by the
kernel.

.PP
The default for this option is \fIdpa\fP.
.SS "-r, --rxq-stats"
.PP
If set, the XDP program will also gather statistics on which receive queue index
each packet was received on. This is displayed in the extended output mode along
with per-CPU data (which, depending on the hardware configuration may or may not
be equivalent).
.SS "-i, --interval <SECONDS>"
.PP
Set the polling interval for collecting all statistics and displaying them to
the output. The unit of interval is in seconds.
.SS "-e, --extended"
.PP
Start xdp-bench in "extended" output mode. If not set, xdp-bench will start in
"terse" mode. The output mode can be switched by hitting C-$\ while the program
is running. See also the \fBOutput Format Description\fP section below.
.SS "-m, --mode"
.PP
Selects the XDP program mode (native or skb). Note that native XDP mode is the
default, and loading the redirect program in skb manner is neither performant,
nor recommended. However, this option is useful if the interface driver lacks
native XDP support, or when simply testing the tool.
.SS "-v, --verbose"
.PP
Enable verbose logging. Supply twice to enable verbose logging from the
underlying \fIlibxdp\fP and \fIlibbpf\fP libraries.
.SS "--version"
.PP
Show the application version and exit.
.SS "-h, --help"
.PP
Display a summary of the available options
.SH "The REDIRECT command"
.PP
In this mode, \fIxdp\-bench\fP sets up packet redirection between the two
interfaces supplied on the command line using the \fIbpf_redirect\fP BPF helper
triggered on packet reception on the ingress interface.

.PP
The syntax for the \fIredirect\fP command is:

.PP
\fIxdp\-bench redirect [options] <ifname_in> <ifname_out>\fP

.PP
Where \fI<ifname_in>\fP is the name of the input interface from where packets will
be redirect to the output interface \fI<ifname_out>\fP.

.PP
The supported options are:
.SS "-l, --load-mode <MODE>"
.PP
Specify which mechanism xdp-bench should use to load (and store) the packet data.
The following modes are available:

.RS
.nf
\fCdpa		- Use traditional Direct Packet Access from the XDP program
load-bytes	- Use the xdp_load_bytes() and xdp_store_bytes() helper functions
\fP
.fi
.RE

.PP
This can be used to benchmark the various packet access modes supported by the
kernel.

.PP
The default for this option is \fIdpa\fP.
.SS "-i, --interval <SECONDS>"
.PP
Set the polling interval for collecting all statistics and displaying them to
the output. The unit of interval is in seconds.
.SS "-s, --stats"
.PP
Enable statistics for successful redirection. This option comes with a per
packet tracing overhead, for recording all successful redirections.
.SS "-e, --extended"
.PP
Start xdp-bench in "extended" output mode. If not set, xdp-bench will start in
"terse" mode. The output mode can be switched by hitting C-$\ while the program
is running. See also the \fBOutput Format Description\fP section below.
.SS "-m, --mode"
.PP
Selects the XDP program mode (native or skb). Note that native XDP mode is the
default, and loading the redirect program in skb manner is neither performant,
nor recommended. However, this option is useful if the interface driver lacks
native XDP support, or when simply testing the tool.
.SS "-v, --verbose"
.PP
Enable verbose logging. Supply twice to enable verbose logging from the
underlying \fIlibxdp\fP and \fIlibbpf\fP libraries.
.SS "--version"
.PP
Show the application version and exit.
.SS "-h, --help"
.PP
Display a summary of the available options
.SH "The REDIRECT-CPU command"
.PP
In this mode, \fIxdp\-bench\fP sets up packet redirection using the
\fIbpf_redirect_map\fP BPF helper triggered on packet reception on the ingress
interface, using a cpumap as its target. Hence, this tool can be used to
redirect packets on an interface from one CPU to another. In addition to this,
the tool then supports redirecting the packet to another output device when it
is processed on the target CPU.

.PP
The syntax for the \fIredirect\-cpu\fP command is:

.PP
\fIxdp\-bench redirect\-cpu [options] <ifname> \-c 0 ... \-c N\fP

.PP
Where \fI<ifname>\fP is the name of the input interface from where packets will be
redirect to the target CPU list specified using \fI\-c\fP.

.PP
The supported options are:
.SS "-c, --cpu <CPU>"
.PP
Specify a possible target CPU index. This option must be passed at least once,
and can be passed multiple times to specify a list of CPUs. Which CPU is chosen
for a given packet depends on the value of the \fI\-\-program\-mode\fP option,
described below.
.SS "-p, --program-mode <MODE>"
.PP
Specify a program that embeds a predefined policy deciding how packets are
redirected to different CPUs. The following options are available:

.RS
.nf
\fCno-touch		- Redirect without touching packet data
touch		- Read packet data before redirecting
round-robin	- Cycle between target CPUs in a round-robin fashion (for each packet)
l4-proto		- Choose the target CPU based on the layer-4 protocol of packet
l4-filter		- Like l4-proto, but drop UDP packets with destination port 9 (used by pktgen)
l4-hash		- Use source and destination IP hashing to pick target CPU
l4-sport		- Use modulo of source port to pick target CPU
l4-dport		- Use modulo of destination port to pick target CPU
\fP
.fi
.RE

.PP
The \fIno\-touch\fP and \fItouch\fP modes always redirect packets to the same CPU (the
first value supplied to \fI\-\-cpu\fP). The \fIround\-robin\fP and \fIl4\-hash\fP modes
distribute packets between all the CPUs supplied as \fI\-\-cpu\fP arguments, while
\fIl4\-proto\fP and \fIl4\-filter\fP send TCP and unrecognised packets to CPU index 0, UDP
packets to CPU index 1 and ICMP packets to CPU index 2 (where the index refers
to the order the actual CPUs are given on the command line).

.PP
The default for this option is \fIl4\-hash\fP.
.SS "-r --remote-action <ACTION>"
.PP
If this option is set, a separate program is installed into the cpumap, which
will be invoked on the remote CPU after the packet is processed there. The
action can be either \fIdrop\fP or \fIpass\fP which will drop the packet or pass it to
the regular networking stack, respectively. Or it can be \fIredirect\fP, which will
cause the packet to be redirected to another interface and transmitted out that
interface on the remote CPU. If this option is set to \fIredirect\fP the target

xdp-bench/xdp-bench.c
#define _GNU_SOURCE
#include <stdio.h>
#include <errno.h>
#include <string.h>

#include "xdp-bench.h"
#include "params.h"

#define PROG_NAME "xdp-bench"

int do_help(__unused const void *cfg, __unused const char *pin_root_path)
{
	fprintf(stderr,
		"Usage: xdp-bench COMMAND [options]\n"
		"\n"
		"COMMAND can be one of:\n"
		"       drop           - Drop all packets on an interface\n"
		"       pass           - Pass all packets to the network stack\n"
		"       tx             - Transmit packets back out on an interface (hairpin forwarding)\n"
		"       redirect       - XDP redirect using the bpf_redirect() helper\n"
		"       redirect-cpu   - XDP CPU redirect using BPF_MAP_TYPE_CPUMAP\n"
		"       redirect-map   - XDP redirect using BPF_MAP_TYPE_DEVMAP\n"
		"       redirect-multi - XDP multi-redirect using BPF_MAP_TYPE_DEVMAP and the BPF_F_BROADCAST flag\n"
		"       xsk-drop       - AF_XDP socket-based drop\n"
		"       xsk-tx         - AF_XDP socket-based hairpin forwarding\n"
		"       help           - show this help message\n"
		"\n"
		"Use 'xdp-bench COMMAND --help' to see options for each command\n");
	return -1;
}


struct enum_val xdp_modes[] = {
       {"native", XDP_MODE_NATIVE},
       {"skb", XDP_MODE_SKB},
       {NULL, 0}
};

struct enum_val basic_program_modes[] = {
       {"no-touch", BASIC_NO_TOUCH},
       {"read-data", BASIC_READ_DATA},
       {"parse-ip", BASIC_PARSE_IPHDR},
       {"swap-macs", BASIC_SWAP_MACS},
       {NULL, 0}
};

struct enum_val basic_load_modes[] = {
       {"dpa", BASIC_LOAD_DPA},
       {"load-bytes", BASIC_LOAD_BYTES},
       {NULL, 0}
};

struct enum_val cpumap_remote_actions[] = {
       {"disabled", ACTION_DISABLED},
       {"drop", ACTION_DROP},
       {"pass", ACTION_PASS},
       {"redirect", ACTION_REDIRECT},
       {NULL, 0}
};

struct enum_val cpumap_program_modes[] = {
       {"no-touch", CPUMAP_NO_TOUCH},
       {"touch", CPUMAP_TOUCH_DATA},
       {"round-robin", CPUMAP_CPU_ROUND_ROBIN},
       {"l4-proto", CPUMAP_CPU_L4_PROTO},
       {"l4-filter", CPUMAP_CPU_L4_PROTO_FILTER},
       {"l4-hash", CPUMAP_CPU_L4_HASH},
       {"l4-sport", CPUMAP_CPU_L4_SPORT},
       {"l4-dport", CPUMAP_CPU_L4_DPORT},
       {NULL, 0}
};

struct enum_val devmap_egress_actions[] = {
       {"forward", DEVMAP_EGRESS_FORWARD },
       {"drop", DEVMAP_EGRESS_DROP },
       {NULL, 0}
};

struct enum_val xsk_program_modes[] = {
       {"rxdrop", XSK_RXDROP},
       {"swap-macs", XSK_SWAP_MACS},
       {NULL, 0}
};

struct enum_val xsk_copy_modes[] = {
       {"auto", XSK_COPY_AUTO},
       {"copy", XSK_COPY_COPY},
       {"zero-copy", XSK_COPY_ZEROCOPY},
       {NULL, 0}
};

struct enum_val xsk_clocks[] = {
       {"MONOTONIC", XSK_CLOCK_MONOTONIC},
       {"REALTIME", XSK_CLOCK_REALTIME},
       {"TAI", XSK_CLOCK_TAI},
       {"BOOTTIME", XSK_CLOCK_BOOTTIME},
       {NULL, 0}
};

struct enum_val xsk_sched_policies[] = {
       {"SCHED_OTHER", XSK_SCHED_OTHER},
       {"SCHED_FIFO", XSK_SCHED_FIFO},
       {NULL, 0}
};

struct prog_option basic_options[] = {
	DEFINE_OPTION("packet-operation", OPT_ENUM, struct basic_opts, program_mode,
		      .short_opt = 'p',
		      .metavar = "<action>",
		      .typearg = basic_program_modes,
		      .help = "Action to take before dropping packet."),
	DEFINE_OPTION("program-mode", OPT_ENUM, struct basic_opts, program_mode,
		      .typearg = basic_program_modes,
		      .hidden = true),
	DEFINE_OPTION("load-mode", OPT_ENUM, struct basic_opts, load_mode,
		      .short_opt = 'l',
                      .metavar = "<mode>",
                      .typearg = basic_load_modes,
		      .help = "How to load (and store) data; default dpa"),
	DEFINE_OPTION("rxq-stats", OPT_BOOL, struct basic_opts, rxq_stats,
		      .short_opt = 'r',
		      .help = "Collect per-RXQ drop statistics"),
	DEFINE_OPTION("interval", OPT_U32, struct basic_opts, interval,
		      .short_opt = 'i',
		      .metavar = "<seconds>",
		      .help = "Polling interval (default 2)"),
	DEFINE_OPTION("extended", OPT_BOOL, struct basic_opts, extended,
		      .short_opt = 'e',
		      .help = "Start running in extended output mode (C^\\ to toggle)"),
	DEFINE_OPTION("xdp-mode", OPT_ENUM, struct basic_opts, mode,
		      .short_opt = 'm',
		      .typearg = xdp_modes,
		      .metavar = "<mode>",
		      .help = "Load XDP program in <mode>; default native"),
	DEFINE_OPTION("dev", OPT_IFNAME, struct basic_opts, iface_in,
		      .positional = true,
		      .metavar = "<ifname>",
		      .required = true,
		      .help = "Load on device <ifname>"),
	END_OPTIONS
};

struct prog_option redirect_basic_options[] = {
	DEFINE_OPTION("load-mode", OPT_ENUM, struct redirect_opts, load_mode,
		      .short_opt = 'l',
                      .metavar = "<mode>",
                      .typearg = basic_load_modes,
		      .help = "How to load (and store) data; default dpa"),
	DEFINE_OPTION("interval", OPT_U32, struct redirect_opts, interval,
		      .short_opt = 'i',
		      .metavar = "<seconds>",
		      .help = "Polling interval (default 2)"),
	DEFINE_OPTION("stats", OPT_BOOL, struct redirect_opts, stats,
		      .short_opt = 's',
		      .help = "Enable statistics for transmitted packets (not just errors)"),
	DEFINE_OPTION("extended", OPT_BOOL, struct redirect_opts, extended,
		      .short_opt = 'e',
		      .help = "Start running in extended output mode (C^\\ to toggle)"),
	DEFINE_OPTION("mode", OPT_ENUM, struct redirect_opts, mode,
		      .short_opt = 'm',
		      .typearg = xdp_modes,
		      .metavar = "<mode>",
		      .help = "Load XDP program in <mode>; default native"),
	DEFINE_OPTION("dev_in", OPT_IFNAME, struct redirect_opts, iface_in,
		      .positional = true,
		      .metavar = "<ifname_in>",
		      .required = true,
		      .help = "Redirect from device <ifname>"),
	DEFINE_OPTION("dev_out", OPT_IFNAME, struct redirect_opts, iface_out,
		      .positional = true,
		      .metavar = "<ifname_out>",
		      .required = true,
		      .help = "Redirect to device <ifname>"),
	END_OPTIONS
};

struct prog_option redirect_cpumap_options[] = {
	DEFINE_OPTION("cpu", OPT_U32_MULTI, struct cpumap_opts, cpus,
		      .short_opt = 'c',
		      .metavar = "<cpu>",
		      .required = true,
		      .help = "Insert CPU <cpu> into CPUMAP (can be specified multiple times)"),
	DEFINE_OPTION("dev", OPT_IFNAME, struct cpumap_opts, iface_in,
		      .positional = true,
		      .metavar = "<ifname>",
		      .required = true,
		      .help = "Run on <ifname>"),
	DEFINE_OPTION("program-mode", OPT_ENUM, struct cpumap_opts, program_mode,
		      .short_opt = 'p',
		      .metavar = "<mode>",
		      .typearg = cpumap_program_modes,
		      .help = "Redirect to CPUs using <mode>. Default l4-hash."),
	DEFINE_OPTION("remote-action", OPT_ENUM, struct cpumap_opts, remote_action,
		      .short_opt = 'r',
		      .metavar = "<action>",
		      .typearg = cpumap_remote_actions,
		      .help = "Perform <action> on the remote CPU. Default disabled."),
	DEFINE_OPTION("redirect-device", OPT_IFNAME, struct cpumap_opts, redir_iface,
		      .short_opt = 'D',
		      .metavar = "<ifname>",
		      .help = "Redirect packets to <ifname> on remote CPU (when --remote-action is 'redirect')"),
	DEFINE_OPTION("qsize", OPT_U32, struct cpumap_opts, qsize,
		      .short_opt = 'q',
		      .metavar = "<packets>",
		      .help = "CPUMAP queue size (default 2048)"),
	DEFINE_OPTION("stress-mode", OPT_BOOL, struct cpumap_opts, stress_mode,
		      .short_opt = 'x',
		      .help = "Stress the kernel CPUMAP setup and teardown code while running"),
	DEFINE_OPTION("interval", OPT_U32, struct cpumap_opts, interval,
		      .short_opt = 'i',
		      .metavar = "<seconds>",
		      .help = "Polling interval (default 2)"),
	DEFINE_OPTION("stats", OPT_BOOL, struct cpumap_opts, stats,
		      .short_opt = 's',
		      .help = "Enable statistics for transmitted packets (not just errors)"),
	DEFINE_OPTION("extended", OPT_BOOL, struct cpumap_opts, extended,
		      .short_opt = 'e',
		      .help = "Start running in extended output mode (C^\\ to toggle)"),
	DEFINE_OPTION("xdp-mode", OPT_ENUM, struct cpumap_opts, mode,
		      .short_opt = 'm',
		      .typearg = xdp_modes,
		      .metavar = "<mode>",
		      .help = "Load XDP program in <mode>; default native"),
	END_OPTIONS
};

struct prog_option redirect_devmap_options[] = {
	DEFINE_OPTION("load-egress", OPT_BOOL, struct devmap_opts, load_egress,
		      .short_opt = 'X',
		      .help = "Load an egress program into the devmap"),
	DEFINE_OPTION("egress-action", OPT_ENUM, struct devmap_opts, egress_action,
		      .short_opt = 'A',
		      .typearg = devmap_egress_actions,
		      .metavar = "<action>",
		      .help = "Egress program <action>. Default is forward"),
	DEFINE_OPTION("interval", OPT_U32, struct devmap_opts, interval,
		      .short_opt = 'i',
		      .metavar = "<seconds>",
		      .help = "Polling interval (default 2)"),
	DEFINE_OPTION("stats", OPT_BOOL, struct devmap_opts, stats,
		      .short_opt = 's',
		      .help = "Enable statistics for transmitted packets (not just errors)"),
	DEFINE_OPTION("extended", OPT_BOOL, struct devmap_opts, extended,
		      .short_opt = 'e',
		      .help = "Start running in extended output mode (C^\\ to toggle)"),
	DEFINE_OPTION("mode", OPT_ENUM, struct devmap_opts, mode,
		      .short_opt = 'm',
		      .typearg = xdp_modes,
		      .metavar = "<mode>",
		      .help = "Load XDP program in <mode>; default native"),
	DEFINE_OPTION("dev_in", OPT_IFNAME, struct devmap_opts, iface_in,
		      .positional = true,
		      .metavar = "<ifname_in>",
		      .required = true,
		      .help = "Redirect from device <ifname>"),
	DEFINE_OPTION("dev_out", OPT_IFNAME, struct devmap_opts, iface_out,
		      .positional = true,
		      .metavar = "<ifname_out>",
		      .required = true,
		      .help = "Redirect to device <ifname>"),
	END_OPTIONS
};

struct prog_option redirect_devmap_multi_options[] = {
	DEFINE_OPTION("load-egress", OPT_BOOL, struct devmap_multi_opts, load_egress,
		      .short_opt = 'X',
		      .help = "Load an egress program into the devmap"),
	DEFINE_OPTION("egress-action", OPT_ENUM, struct devmap_multi_opts, egress_action,
		      .short_opt = 'A',
		      .typearg = devmap_egress_actions,
		      .metavar = "<action>",
		      .help = "Egress program <action>. Default is forward"),
	DEFINE_OPTION("interval", OPT_U32, struct devmap_multi_opts, interval,
		      .short_opt = 'i',
		      .metavar = "<seconds>",
		      .help = "Polling interval (default 2)"),
	DEFINE_OPTION("stats", OPT_BOOL, struct devmap_multi_opts, stats,
		      .short_opt = 's',
		      .help = "Enable statistics for transmitted packets (not just errors)"),
	DEFINE_OPTION("extended", OPT_BOOL, struct devmap_multi_opts, extended,
		      .short_opt = 'e',
		      .help = "Start running in extended output mode (C^\\ to toggle)"),
	DEFINE_OPTION("mode", OPT_ENUM, struct devmap_multi_opts, mode,
		      .short_opt = 'm',
		      .typearg = xdp_modes,
		      .metavar = "<mode>",
		      .help = "Load XDP program in <mode>; default native"),
	DEFINE_OPTION("devs", OPT_IFNAME_MULTI, struct devmap_multi_opts, ifaces,
		      .positional = true,
		      .metavar = "<ifname...>",
		      .min_num = 2,
		      .max_num = MAX_IFACE_NUM,
		      .required = true,
		      .help = "Redirect from and to devices <ifname...>"),
	END_OPTIONS
};

struct prog_option xsk_options[] = {
	DEFINE_OPTION("queue", OPT_U32, struct xsk_opts, queue_idx,
		      .short_opt = 'q',
		      .metavar = "<queue>",
		      .help = "Queue index to use (default 0)"),
	DEFINE_OPTION("interval", OPT_U32, struct xsk_opts, interval,
		      .short_opt = 'i',
		      .metavar = "<seconds>",
		      .help = "Statistics update interval (default 2)"),
	DEFINE_OPTION("frame-size", OPT_U32, struct xsk_opts, frame_size,
		      .short_opt = 'f',
		      .metavar = "<size>",
		      .help = "Data frame size (must be a power of two in aligned mode); (default 4096)"),
	DEFINE_OPTION("duration", OPT_U32, struct xsk_opts, duration,
		      .short_opt = 'd',
		      .metavar = "<seconds>",
		      .help = "Duration to run; default 0 (forever)"),
	DEFINE_OPTION("batch-size", OPT_U32, struct xsk_opts, batch_size,
		      .short_opt = 'b',
		      .metavar = "<packets>",
		      .help = "Batch size for receive loop; default 64"),
	DEFINE_OPTION("irq-string", OPT_STRING, struct xsk_opts, irq_string,
		      .short_opt = 'I',
		      .metavar = "<irq-string>",
		      .help = "Display driver interrupt statistics for interface associated with <irq-string>"),
	DEFINE_OPTION("poll", OPT_BOOL, struct xsk_opts, use_poll,
		      .short_opt = 'p',
		      .help = "Use poll syscall"),
	DEFINE_OPTION("no-need-wakeup", OPT_BOOL, struct xsk_opts, no_need_wakeup,
		      .short_opt = 'm',
		      .help = "Turn off use of driver need wakeup flag"),
	DEFINE_OPTION("unaligned", OPT_BOOL, struct xsk_opts, unaligned,
		      .short_opt = 'u',
		      .help = "Enable unaligned chunk placement"),
	DEFINE_OPTION("shared-umem", OPT_BOOL, struct xsk_opts, shared_umem,
		      .short_opt = 'M',
		      .help = "Enable XDP_SHARED_UMEM across multiple sockets"),
	DEFINE_OPTION("extra-stats", OPT_BOOL, struct xsk_opts, extra_stats,
		      .short_opt = 'x',
		      .help = "Display extra statistics"),
	DEFINE_OPTION("quiet", OPT_BOOL, struct xsk_opts, quiet,
		      .short_opt = 'Q',
		      .help = "Do not display any stats"),
	DEFINE_OPTION("app-stats", OPT_BOOL, struct xsk_opts, app_stats,
		      .short_opt = 'a',
		      .help = "Display application (syscall) statistics"),
	DEFINE_OPTION("busy-poll", OPT_BOOL, struct xsk_opts, busy_poll,
		      .short_opt = 'B',
		      .help = "Enable busy polling"),
	DEFINE_OPTION("frags", OPT_BOOL, struct xsk_opts, frags,
		      .short_opt = 'F',
		      .help = "Enable frags (multi-buffer) support"),
	DEFINE_OPTION("copy-mode", OPT_ENUM, struct xsk_opts, copy_mode,
		      .short_opt = 'C',
		      .typearg = xsk_copy_modes,
		      .metavar = "<mode>",
		      .help = "Use <mode> for copying data packets to userspace; default auto"),
	DEFINE_OPTION("clock", OPT_ENUM, struct xsk_opts, clock,
		      .short_opt = 'w',
		      .typearg = xsk_clocks,
		      .metavar = "<clock>",
		      .help = "Clock name to use; default MONOTONIC"),
	DEFINE_OPTION("policy", OPT_ENUM, struct xsk_opts, sched_policy,
		      .short_opt = 'W',
		      .typearg = xsk_sched_policies,
		      .metavar = "<policy>",
		      .help = "Scheduler policy; default SCHED_OTHER"),
	DEFINE_OPTION("schpri", OPT_U32, struct xsk_opts, sched_prio,
		      .short_opt = 'U',
		      .metavar = "<priority>",
		      .help = "Scheduler priority; default 0"),
	DEFINE_OPTION("attach-mode", OPT_ENUM, struct xsk_opts, attach_mode,
		      .short_opt = 'A',
		      .typearg = xdp_modes,
		      .metavar = "<mode>",
		      .help = "Load XDP program in <mode>; default native"),
	DEFINE_OPTION("dev", OPT_IFNAME, struct xsk_opts, iface,
		      .positional = true,
		      .metavar = "<ifname>",
		      .required = true,
		      .help = "Load on device <ifname>"),
	END_OPTIONS
};

static const struct prog_command cmds[] = {
	{ .name = "drop",
	  .func = do_drop,
	  .options = basic_options,
	  .default_cfg = &defaults_drop,
	  .doc = "Drop all packets on an interface" },
	{ .name = "pass",
	  .func = do_pass,
	  .options = basic_options,
	  .default_cfg = &defaults_pass,
	  .doc = "Pass all packets to the network stack" },
	{ .name = "tx",
	  .func = do_tx,
	  .options = basic_options,
	  .default_cfg = &defaults_tx,
	  .doc = "Transmit packets back out an interface (hairpin forwarding)" },
	DEFINE_COMMAND_NAME("redirect", redirect_basic,
			    "XDP redirect using the bpf_redirect() helper"),
	DEFINE_COMMAND_NAME("redirect-cpu", redirect_cpumap,
			    "XDP CPU redirect using BPF_MAP_TYPE_CPUMAP"),
	DEFINE_COMMAND_NAME("redirect-map", redirect_devmap,
			    "XDP redirect using BPF_MAP_TYPE_DEVMAP"),
	DEFINE_COMMAND_NAME(
		"redirect-multi", redirect_devmap_multi,
		"XDP multi-redirect using BPF_MAP_TYPE_DEVMAP and the BPF_F_BROADCAST flag"),
	{ .name = "xsk-drop",
	  .func = do_xsk_drop,
	  .options = xsk_options,
	  .default_cfg = &defaults_xsk,
	  .doc = "AF_XDP-based packet drop" },
	{ .name = "xsk-tx",
	  .func = do_xsk_tx,
	  .options = xsk_options,
	  .default_cfg = &defaults_xsk,
	  .doc = "AF_XDP-based transmit back out an interface (hairpin forwarding)" },
	{ .name = "help", .func = do_help, .no_cfg = true },
	END_COMMANDS
};

union all_opts {
	struct basic_opts basic;
	struct cpumap_opts cpumap;
	struct devmap_opts devmap;
	struct devmap_multi_opts devmap_multi;
	struct xsk_opts xsk;
};

int main(int argc, char **argv)
{
	if (argc > 1)
		return dispatch_commands(argv[1], argc - 1, argv + 1, cmds,
					 sizeof(union all_opts), PROG_NAME, false);

	return do_help(NULL, NULL);
}
